wip-0003.md

```diff
@@ -0,0 +1,136 @@
+
+BLS signatures allow us to aggregate signatures provided by a set of group members. In the case of Witnet, those members will be the ARS, as the consensus of the Witnet chain is highly dependant on the reputation of each of the peers. The overall aggregated signature can be validated in the target chain in a single verification. However, a BLS verification is approximately 10 times more costly than a regular ECDSA verification. Thus, BLS gives us a big advantage for bridging with other chains, but little advantage when using it to sign transactions internally in the _Witnet network_.
+
+In consequence, we need to derive a way to make both BLS and ECDSA algorithms co-exist in Witnet. This proposal introduces different solutions to such problem.
+
+## Requirements
+As already said before, we need to maintain a state about the ARS in the block headers that future blocks can be validated upon. Such information will be stored as a merkle root of all the identities and their position in the ARS.
+
+Note that reputation score is currently associated to the hash of the Secp256K1 public key. However, ARS members in the ARS merkle root should be identifiable by their BLS public key, as this will be used to verify their aggregated signature.
+
+Additionally, addresses and transaction outputs are also associated with the hash of the Secp256k1 public key. Given the clear overhead of implementing a BLS verification compared to a ECDSA verification, transaction outputs should keep being spendable by secp256k1 private key holders.
+
+## Proposal: A mapping between Secp256 and BN256 public keys.
+
+### Specification
+Our first proposal is to make every node store a mapping with the relationship between a *Secp256k1* and the *BN256* key.
+
+#### Information storage
+
+**`secp256k1bn256Map` mapping.** A new mapping to be stored in memory that is updated every time a new key is proposed. This mapping does not need to be stored on-chain, but can be constructed by information that is on-chain. This can be implemented as an additional mapping to be stored in the chain manager.
+
+``` rust
+type secp256k1bn256Map = HashMap<PublicKeyHash, Bn256PublicKey>;
+```
+
+#### Commitment Transactions
+
+**`bn256_pub_key` field.** One new field in the commitment transaction body (`CommitTransactionBody`) data structure `bn256_pub_key` is defined:
+
+```protobuf
+message CommitTransactionBody {
+    Hash dr_pointer = 1;
+    Hash commitment = 2;
+    DataRequestEligibilityClaim proof = 3;
+    repeated Input inputs = 4;
+    repeated ValueTransferOutput outputs = 5;
+    Bn256PublicKey bn256_pub_key = 6;
```

Comment:
I suggest using extended naming as I think we are doing in the rest of `witnet-rust` code:

```suggestion
    Bn256PublicKey bn256_public_key = 6;
```